[FLINK-3921] StringParser encoding

greghogan opened the pull request:
This further modifies #2060:
- save charset as String which is serializable
- remove unused methods
- annotate new CsvReader methods as @PublicEvolving

This can be squashed into the first commit after review.

fhueske commented:
Hi @greghogan, thanks for working on this. The changes look good. 
However, I think we need to propagate the configured charset also to the `DelimitedInputFormat`. `DelimitedInputFormat` splits the file content into lines based on a row delimiter which is always encoded as `UTF-8`. Each line is subsequently parsed by the logic of the `CsvInputFormat`.

greghogan commented:
@fhueske, please review the additional commits.

fhueske commented on flink-core/src/main/java/org/apache/flink/api/common/io/DelimitedInputFormat.java:
```diff
@@ -182,8 +186,37 @@ protected DelimitedInputFormat(Path filePath, Configuration configuration) {
+	 * Set the name of the character set used for the row delimiter, field
+	 * delimiter, comment string, and {@link FieldParser}.
+	 *
+	 * The delimeters and comment string are interpreted when set. Each
```
Can we change the encoding of the current delimiter using the old charset and the new charset?
That would also allow to set a new charset without the need to set `\n` again to have a newline delimiter.

fhueske replied:
`GenericCsvInputFormat` would need to override this method and call `super.setCharset()` to propagate the change.

What do you think @greghogan?

fhueske replied:
Btw., the `DelimitedInputFormat` does not have a comment string or field parsers.

fhueske commented on flink-core/src/test/java/org/apache/flink/api/common/io/DelimitedInputFormatTest.java:
```diff
@@ -200,6 +200,30 @@ public void testReadCustomDelimiter() {
 	}
 
+	@Test
+	public void testReadCustomDelimiterWithCharset() throws IOException {
```
This tests the `DelimitedInputFormat` with default charset. 
Can you add a test with a different charset as well?

greghogan commented:
@fhueske, I refactored `DelimitedInputFormat` and `GenericCsvInputFormat` to reinterpret the fields whenever the charset encoding changes, and also updated the test per your review.

fhueske commented:
Thanks for the update @greghogan.
PR looks good to merge. 

Something that came to my mind. Some of the FieldParsers such as `DoubleParser` create a String before converting it into a target value. Shouldn't we use the correct charset in these cases as well?

greghogan commented:
@fhueske, the `FieldParser`s look to use two methods to parse their fields. `DoubleParser` uses `FieldParser.nextStringEndPos`, recently added by Timo, to create a `String` from the byte array which can be easily interpreted with a non-default `Charset`. The `String` is then parsed by `Double.parseDouble`. The `IntParser` processes one character at a time ad-hoc.

Should we adopt the first method of converting bytes -> String -> parse for all `FieldParser`s? This could be done in a separate ticket.

fhueske commented:
The byte-level implementation as in `IntParser` was originally used to avoid String object instances. Such an implementation was not possible with Double because it led to very imprecise results. Therefore, we chose the String approach there. You are of course right, that we would need to use the String approach as well if a charset is used that is not compatible with the current byte-level parsing.

IMO, it makes sense to open a JIRA for that and solving this as a follow up to this issue.
Feel free to merge this PR.

fhueske commented:
Merging (+ closing PR #2060 which is included in this one)

greghogan commented:
I had planned to squash these into the original commit (#2060), but whatever you think is best.

fhueske commented:
I thought, I'll merge the original commit and squash yours on top. OK?

greghogan commented:
I'm not quite following but I think we have the same idea.
